python/ray/tune/syncer.py

```diff
@@ -142,395 +79,408 @@ class SyncConfig:
-        sync_function: Function for syncing the local_dir to
-            remote_dir. If string, then it must be a string template for
-            syncer to run. If True or not provided, it defaults rsync
-            (if available) or otherwise remote-task based syncing. If
-            False, a noop Syncer is returned.
-    """
-    if sync_function == "auto":
-        sync_function = None  # Auto-detect
-
-    key = (local_dir, remote_dir)
-    if key in _syncers:
-        # Get cached syncer
-        return _syncers[key]
-    elif isclass(sync_function) and issubclass(sync_function, Syncer):
-        # Type[Syncer]
-        _syncers[key] = sync_function(local_dir, remote_dir, None)
-        return _syncers[key]
-    elif not remote_dir or sync_function is False:
-        # Do not sync trials if no remote dir specified or syncer=False
-        sync_client = NOOP
-    elif sync_function and sync_function is not True:
-        # String or callable (for function syncers)
-        sync_client = get_sync_client(sync_function)
-    else:
-        # sync_function=True or sync_function=None --> default
-        rsync_function_str = get_rsync_template_if_available()
-        if rsync_function_str:
-            sync_client = CommandBasedClient(rsync_function_str, rsync_function_str)
-            sync_client.set_logdir(local_dir)
-        else:
-            sync_client = RemoteTaskClient()
+    def _local_trial_logdir(self, trial: "Trial"):
+        return trial.logdir
 
-    _syncers[key] = NodeSyncer(local_dir, remote_dir, sync_client)
-    return _syncers[key]
+    def _remote_trial_logdir(self, trial: "Trial"):
+        return trial.logdir
 
+    def _sync_trial_dir(
+        self, trial: "Trial", force: bool = False, wait: bool = True
+    ) -> bool:
+        if not self._enabled or trial.uses_cloud_checkpointing:
+            return False
 
-@DeveloperAPI
-class SyncerCallback(Callback):
-    def __init__(self, sync_function: Optional[Union[bool, Callable]]):
-        self._sync_function = sync_function
-        self._syncers: Dict["Trial", NodeSyncer] = {}
-
-    def _get_trial_syncer(self, trial: "Trial"):
-        if trial not in self._syncers:
-            self._syncers[trial] = self._create_trial_syncer(trial)
-        return self._syncers[trial]
-
-    def _create_trial_syncer(self, trial: "Trial"):
-        return get_node_syncer(
-            trial.logdir, remote_dir=trial.logdir, sync_function=self._sync_function
-        )
+        sync_process = self._get_trial_sync_process(trial)
 
-    def _remove_trial_syncer(self, trial: "Trial"):
-        self._syncers.pop(trial, None)
+        if not force and not self._should_sync(trial):
+            return False
 
-    def _sync_trial_checkpoint(self, trial: "Trial", checkpoint: _TrackedCheckpoint):
-        if checkpoint.storage_mode == CheckpointStorage.MEMORY:
-            return
+        if NODE_IP in trial.last_result:
+            source_ip = trial.last_result[NODE_IP]
```

Comment:
Off the topic as again I know this is current implementation. But I am wondering if this should really come from TrackedCheckpoint? What if we are talking about syncing checkpoint if distributed HuggingFace case? Where the ckpt is at is different than the trial.last_result[NODE_IP].

Reply:
If I understand correctly we will still have a metadata checkpoint at the last_result[NODE_IP] location that then points to the actual location. Also the checkpoint manager does `self.node_ip = node_ip or self.metrics.get(NODE_IP, None)` so I'm wondering if these are actually ever out of sync.